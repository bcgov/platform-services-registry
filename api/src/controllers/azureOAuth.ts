import axios from "axios";
import { logger } from "@bcgov/common-nodejs-utils";
// import { Response } from "express";
import qs from "qs";
import { Request, Response } from "express";

export const fetchAzureAccessToken = async (req: Request, res: Response): Promise<void> => {
  try {
    const options = {
      Host: "login.microsoftonline.com",
      client_id: `${process.env.AZURE_CLIENT_ID}`,
      scope: "https://graph.microsoft.com/.default",
      client_secret: process.env.AZURE_CLIENT_SECRET,
      grant_type: "client_credentials",
    };
    const url: string = `https://login.microsoftonline.com/${process.env.AZURE_TENANT_ID}/oauth2/v2.0/token`;
    axios
      .post(url, qs.stringify(options))
      .then((response) => {
        if (!response.data.access_token) {
          throw Error(
            "a result was returned from Microsoft's token endpoint, but it doesn't seem to have an access token included"
          );
        }
        res.status(200).json(response.data.access_token);
      })
      .catch((error) => {
        logger.info(`error: ${error}`);
        res.status(500);
      });
  } catch (err) {
    logger.error(`err = ${err.message}`);
  }
};

export default fetchAzureAccessToken;
