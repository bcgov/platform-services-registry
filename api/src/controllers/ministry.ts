//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from "@bcgov/common-nodejs-utils";
import { Request, Response } from "express";
import DataManager from "../db";
import shared from "../libs/shared";

const dm = new DataManager(shared.pgPool);

const fetchMinistrySponsors = async (
  req: Request,
  res: Response
): Promise<void> => {
  const { MinistryModel } = dm;
  try {
    const results = await MinistryModel.findAll();

    res.status(200).json(results);
  } catch (err) {
    const message = "Unable fetch ministries";
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export default fetchMinistrySponsors;
