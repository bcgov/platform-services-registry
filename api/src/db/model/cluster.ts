//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { CommonFields, Model } from './model';

export interface Cluster extends CommonFields {
  name: string;
  description: string;
  disasterRecovery: boolean;
  onPrem: boolean;
  onHardware: boolean;
  isDefault: boolean;
  isProd: boolean;
}

export default class ClusterModel extends Model {
  table: string = 'ref_cluster';
  requiredFields: string[] = [
    'name',
    'description',
    'disasterRecovery',
    'onPrem',
    'onHardware',
    'isProd'
  ];
  pool: Pool;

  constructor(pool: any) {
    super();
    this.pool = pool;
  }

  async create(data: Cluster): Promise<Cluster> {
    const query = {
      text: `
        INSERT INTO ${this.table}
          (name, description, disaster_recovery, on_prem, on_hardware)
          VALUES ($1, $2, $3, $4, $5) RETURNING *;`,
      values: [
        data.name,
        data.description,
        data.disasterRecovery,
        data.onPrem,
        data.onHardware,
      ],
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to create cluster`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findDefault(): Promise<Cluster> {
    const query = {
      text: `
        SELECT * FROM ${this.table}
          WHERE is_default = $1 AND archived = false;
      `,
      values: [
        true,
      ],
    };

    try {
      const results = await this.runQuery(query);

      return results.pop();
    } catch (err) {
      const message = `Unable to lookup default cluster`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findByName(name: string): Promise<Cluster> {
    const query = {
      text: `
        SELECT * FROM ${this.table} WHERE name = $1;
      `,
      values: [
        name,
      ],
    };

    try {
      const results = await this.runQuery(query);

      return results.pop();
    } catch (err) {
      const message = `Unable to lookup cluster by name ${name}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findAllProdReady(): Promise<Cluster[]> {
    const query = {
      text: `
        SELECT * FROM ${this.table} WHERE is_prod = $1 AND archived = false;
      `,
      values: [
        true,
      ],
    };

    try {
      return await this.runQuery(query);
    } catch (err) {
      const message = `Unable to find all production ready clusters`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async update(clusterId: number, data: Cluster): Promise<Cluster> {
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            name = $1, description = $2, disaster_recovery = $3, on_prem = $4,
            on_hardware = $5,
          WHERE id = $6
          RETURNING *;`,
      values: [
        data.name,
        data.description,
        data.disasterRecovery,
        data.onPrem,
        data.onHardware,
        clusterId,
      ],
    };

    try {
      const results = await this.runQuery(query);

      return results.pop();
    } catch (err) {
      const message = `Unable to update cluster ${clusterId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async delete(clusterId: number): Promise<Cluster> {
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            archived = true
          WHERE id = $1
          RETURNING *;
      `,
      values: [
        clusterId,
      ],
    };

    try {
      const results = await this.runQuery(query);

      return results.pop();
    } catch (err) {
      const message = `Unable to archive cluster`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }
}
