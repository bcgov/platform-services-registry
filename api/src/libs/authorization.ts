//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { errorWithCode } from '@bcgov/common-nodejs-utils';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import shared from '../libs/shared';

const dm = new DataManager(shared.pgPool);

export const getAuthorization = async(profileId: number, user: any, projectDetails?:any, projectContacts?:any): Promise<boolean | Error> => {
  const { ProfileModel, ContactModel } = dm;

  if (!projectDetails){
    projectDetails = await ProfileModel.findById(Number(profileId));
  }

  if (!projectContacts){
    projectContacts = await ContactModel.findForProject(Number(profileId));
  }

  const authorizedEmails = projectContacts.map(contact => contact.email);

  if (!(
    user.id === projectDetails.userId
    || authorizedEmails.includes(user.email)
    || user.roles.includes(USER_ROLES.ADMINISTRATOR)
  )) {
    return errorWithCode('Unauthorized Access', 401);
  }

  return true;
}
