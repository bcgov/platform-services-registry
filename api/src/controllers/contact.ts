//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-15.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;

  const rv = validateObjProps(ContactModel.requiredFields, body);
  if (rv) {
    throw rv;
  }

  try {
    const result = await ContactModel.create(body);

    res.status(201).json(result);
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { contactId } = params;
  const {
    firstName,
    lastName,
    email,
    githubId,
  } = body;

  try {
    const record = await ContactModel.findById(contactId);
    const aBody = {
      userId: record.userId,
      firstName,
      lastName,
      email,
      githubId,
      roleId: record.roleId,
    };

    const rv = validateObjProps(ContactModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ContactModel.update(contactId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update contact ID ${contactId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const requestContactEdit = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
