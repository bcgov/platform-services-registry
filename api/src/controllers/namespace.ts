//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;
  const aBody = { ...body, profileId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    const results = await NamespaceModel.create(aBody);

    res.status(201).json(results);
  } catch (err) {
    const message = 'Unable create new namespace';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespaces = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;

  try {
    const results = await NamespaceModel.findForProfile(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { namespaceId } = params;

  try {
    const results = await NamespaceModel.findById(Number(namespaceId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace with ID ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;
  const { name, clusterId } = body;
  const aBody = { name, profileId, clusterId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    await NamespaceModel.update(namespaceId, aBody);

    res.status(204).end();
  } catch (err) {
    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;

  try {
    await NamespaceModel.delete(namespaceId);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to archive profile ${profileId} namespace ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
