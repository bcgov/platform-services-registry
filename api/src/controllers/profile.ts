//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import { generateNamespacePrefix } from '../db/utils';
import { AuthenticatedUser } from '../libs/authmware';
import shared from '../libs/shared';
import { isNotAuthorized, validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
  const { ProfileModel } = dm;
  const attempts = 3;
  const candidates = Array.from({ length: attempts }, () => generateNamespacePrefix())
  const promises = candidates.map(c => ProfileModel.isNamespacePrefixUnique(c));

  try {
    const results = await Promise.all(promises);
    const values = results.map((cur, idx) => {
      if (cur) {
        return candidates[idx];
      }

      return;
    }).filter(v => typeof v !== 'undefined');

    return values.pop();
  } catch (err) {
    logger.error(err.message);
    return;
  }
}

export const fetchAllProjectProfiles = async (
  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
  const { ProfileModel } = dm;

  try {
    let results;
    if (user.roles.includes(USER_ROLES.ADMINISTRATOR)) {
      results = await ProfileModel.findAll();
    } else {
      results = await ProfileModel.findProfilesByUserId(user.id);
    }

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    res.status(200).json(record);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const createProjectProfile = async (
  { body, user }: { body: any, user: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const data = { ...body, userId: user.id };

  // User cannot set the namespace prefix. If it exists, this
  // overwrites it with a place holder value. It will be replaced
  // with the actual value further on.
  const rv = validateObjProps(ProfileModel.requiredFields, {
    ...data,
    namespacePrefix: 'placeholder',
  });

  if (rv) {
    throw rv;
  }

  try {
    const namespacePrefix = await uniqueNamespacePrefix();

    if (!namespacePrefix) {
      throw errorWithCode(500, 'Unable to generate unique namespace prefix');
    }

    const results = await ProfileModel.create({ ...data, namespacePrefix });

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProjectProfile = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const {
    name,
    description,
    categoryId,
    busOrgId,
    prioritySystem,
    notificationEmail,
    notificationSms,
    notificationMsTeams,
    paymentBambora,
    paymentPayBc,
    fileTransfer,
    fileStorage,
    geoMappingWeb,
    geoMappingLocation,
    schedulingCalendar,
    schedulingAppointments,
    idmSiteMinder,
    idmKeycloak,
    idmActiveDir,
    other,
  } = body;

  try {
    const record = await ProfileModel.findById(profileId);
    const aBody = {
      name,
      description,
      categoryId,
      busOrgId,
      prioritySystem,
      userId: record.userId,
      namespacePrefix: record.namespacePrefix,
      notificationEmail,
      notificationSms,
      notificationMsTeams,
      paymentBambora,
      paymentPayBc,
      fileTransfer,
      fileStorage,
      geoMappingWeb,
      geoMappingLocation,
      schedulingCalendar,
      schedulingAppointments,
      idmSiteMinder,
      idmKeycloak,
      idmActiveDir,
      other,
    };

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const rv = validateObjProps(ProfileModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ProfileModel.update(profileId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(profileId);

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    await ProfileModel.delete(profileId);

    res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = 'Unable to archive project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const addContactToProfile = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId, contactId } = params;

  try {
    await ProfileModel.addContactToProfile(Number(profileId), Number(contactId));

    res.status(202).end();
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileContacts = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ContactModel, ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const results = await ContactModel.findForProject(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch profile contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
