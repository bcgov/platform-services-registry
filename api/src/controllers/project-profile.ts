//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import { generateNamespacePrefix } from '../db/utils';
import { AuthenticatedUser } from '../libs/authmware';
import { getAuthorization } from '../libs/authorization';
import { getDefaultCluster } from '../libs/profile';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
  const { ProfileModel } = dm;
  const attempts = 3;
  const candidates = Array.from({ length: attempts }, () => generateNamespacePrefix())
  const promises = candidates.map(c => ProfileModel.isNamespacePrefixUnique(c));

  try {
    const results = await Promise.all(promises);
    const values = results.map((cur, idx) => {
      if (cur) {
        return candidates[idx];
      }

      return;
    }).filter(v => typeof v !== 'undefined');

    return values.pop();
  } catch (err) {
    logger.error(err.message);
    return;
  }
}

export const fetchAllProjectProfiles = async (
  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
  const { ProfileModel } = dm;

  try {
    let results;
    if (user.roles.includes(USER_ROLES.ADMINISTRATOR)) {
      results = await ProfileModel.findAll();
    } else {
      results = await ProfileModel.findProfilesByUserIdOrEmail(user.id, user.email);
    }

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const projectDetails = await ProfileModel.findById(Number(profileId));

    const isAuthorized = getAuthorization(profileId, user, projectDetails);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    res.status(200).json(projectDetails);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const createProjectProfile = async (
  { body, user }: { body: any, user: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const data = { ...body, userId: user.id };

  // User cannot set the namespace prefix OR primary cluster name (current default cluster).
  // If it exists, this overwrites it with a place holder value. It will be replaced
  // with the actual value further on.
  const rv = validateObjProps(ProfileModel.requiredFields, {
    ...data,
    namespacePrefix: 'placeholder',
    primaryClusterName: 'placeholder',
  });

  if (rv) {
    throw rv;
  }

  try {
    const namespacePrefix = await uniqueNamespacePrefix();
    if (!namespacePrefix) {
      throw errorWithCode(500, 'Unable to generate unique namespace prefix');
    }

    const defaultCluster = await getDefaultCluster();
    if (!defaultCluster) {
      throw errorWithCode(500, 'Unable to set primary cluster id based on default cluster');
    }

    const results = await ProfileModel.create({ ...data, namespacePrefix, primaryClusterName: defaultCluster.name });

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProjectProfile = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const {
    name,
    description,
    busOrgId,
    prioritySystem,
    notificationEmail,
    notificationSms,
    notificationMsTeams,
    paymentBambora,
    paymentPayBc,
    fileTransfer,
    fileStorage,
    geoMappingWeb,
    geoMappingLocation,
    schedulingCalendar,
    schedulingAppointments,
    idmSiteMinder,
    idmKeycloak,
    idmActiveDir,
    other,
    migratingLicenseplate,
  } = body;

  try {
    const currentProjectDetails = await ProfileModel.findById(profileId);
    const aBody = {
      name,
      description,
      busOrgId,
      prioritySystem,
      userId: currentProjectDetails.userId,
      namespacePrefix: currentProjectDetails.namespacePrefix,
      notificationEmail,
      notificationSms,
      notificationMsTeams,
      paymentBambora,
      paymentPayBc,
      fileTransfer,
      fileStorage,
      geoMappingWeb,
      geoMappingLocation,
      schedulingCalendar,
      schedulingAppointments,
      idmSiteMinder,
      idmKeycloak,
      idmActiveDir,
      other,
      primaryClusterName: currentProjectDetails.primaryClusterName,
      migratingLicenseplate,
    };

    const isAuthorized = getAuthorization(profileId, user, currentProjectDetails);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    const rv = validateObjProps(ProfileModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ProfileModel.update(profileId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const isAuthorized = getAuthorization(profileId, user);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    await ProfileModel.delete(profileId);

    res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = 'Unable to archive project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
