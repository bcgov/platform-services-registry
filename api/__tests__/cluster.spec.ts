//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import { fetchClusters } from '../src/controllers/cluster';
import ClusterModel from '../src/db/model/cluster';
import { AccessFlag } from '../src/libs/authorization';
import FauxExpress from './src/fauxexpress';

const p0 = path.join(__dirname, 'fixtures/select-default-cluster.json');
const selectCluster = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/get-authenticated-user.json');
const authenticatedUser = JSON.parse(fs.readFileSync(p1, 'utf8'));

const client = new Pool().connect();

describe('Project-profile event handlers', () => {
  let ex;

  beforeEach(() => {
    jest.clearAllMocks();

    ex = new FauxExpress();
  });


  it('All clusters are returned', async () => {
    const req = {
      user: { ...authenticatedUser, accessFlags: [AccessFlag.ProvisionOnTestCluster,] },
    };
    client.query.mockReturnValueOnce({ rows: selectCluster });

    // @ts-ignore
    await fetchClusters(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch all clusters should throw', async () => {
    const req = {};
    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(fetchClusters(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('Prod-ready clusters are returned to non-authorized user', async () => {
    const findAllProdReady = ClusterModel.prototype.findAllProdReady = jest.fn().mockResolvedValue({ rows: ['stub'] });
    const req = {
      user: authenticatedUser,
    };

    // @ts-ignore
    await fetchClusters(req, ex.res);
    expect(findAllProdReady).toHaveBeenCalledTimes(1);
  });

  it('Both test and prod-ready clusters are returned to authorized user', async () => {
    const findAll = ClusterModel.prototype.findAll = jest.fn().mockResolvedValue({ rows: ['stub'] });
    const req = {
      user: { ...authenticatedUser, accessFlags: [AccessFlag.ProvisionOnTestCluster,] },
    };

    // @ts-ignore
    await fetchClusters(req, ex.res);
    expect(findAll).toHaveBeenCalledTimes(1);
  });
});
