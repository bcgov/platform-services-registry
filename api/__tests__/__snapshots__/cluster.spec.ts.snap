// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project-profile event handlers All clusters are returned 1`] = `
Array [
  Array [
    Object {
      "text": "
      SELECT * FROM ref_cluster
        WHERE archived = false;
      ",
    },
  ],
]
`;

exports[`Project-profile event handlers All clusters are returned 2`] = `200`;

exports[`Project-profile event handlers All clusters are returned 3`] = `
Array [
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.304Z",
    "description": "Silver cluster",
    "disasterRecovery": false,
    "displayName": "Silver Kamloops",
    "id": 1,
    "isDefault": true,
    "isProd": true,
    "name": "silver",
    "onPrem": true,
    "updatedAt": "2020-09-10T18:14:13.304Z",
  },
]
`;

exports[`Project-profile event handlers Fetch all clusters should throw 1`] = `"Unable to fetch clusters"`;

exports[`Project-profile event handlers Fetch all clusters should throw 2`] = `Array []`;
